# **Data Preparation** — `01_data_preparation.ipynb`

## **Objective**

This notebook prepares the NYC taxi dataset for further analysis.

The objectives are to:

* understand the structure and grain of the dataset;
* inspect the variables and their data types;
* identify missing and duplicated records;
* investigate categorical and numerical values;
* check for potential data-quality issues and logical inconsistencies;
* apply justified data-cleaning and data-type conversion steps;
* validate the resulting dataset;
* save an analysis-ready version for the following notebooks.

In [1]:
# importing pandas for Data manipulation
import pandas as pd

In [2]:
# loading the raw nyc taxi Dataset
df = pd.read_csv("../data/raw/aggregated_nyc_yellow_taxi_2024.csv")

In [3]:
# Data head preview 
df.head(3)

,date,hour,passenger_count,PU_Borough,DO_Borough,payment_type,trip_count,trip_distance_sum,duration_sum,fare_amount_sum,extra_sum,mta_tax_sum,tip_amount_sum,tolls_amount_sum,improvement_surcharge_sum,congestion_surcharge_sum,airport_fee_sum,total_amount_sum
0,2024-01-01,00:00,1.0,Bronx,Bronx,2,2,1.60,9.216667,13.0,2.0,1.0,0.00,0.0,2.0,0.0,0.0,18.00
1,2024-01-01,00:00,1.0,Bronx,Manhattan,2,1,7.50,18.666667,32.4,1.0,0.5,0.00,0.0,1.0,0.0,0.0,34.90
2,2024-01-01,00:00,1.0,Brooklyn,Brooklyn,1,6,10.28,42.750000,60.7,6.0,3.0,11.65,0.0,6.0,0.0,0.0,87.35


In [4]:
# index Inspection 
print("Index unique:", df.index.is_unique)
print("Index sequential:", df.index.equals(pd.RangeIndex(start=0, stop=len(df))))

Index unique: True
Index sequential: True


In [5]:
# shape Inspection
df.shape

(921373, 18)

In [6]:
# missing values Inspection
df.isna().sum().sum()

np.int64(0)

In [7]:
# duplicated values Inspection
df.duplicated().sum()

np.int64(0)

In [8]:
# statistical Inspection
df.describe(include="all").T[["unique", "min", "mean", "50%", "max"]].assign(unique=df.nunique(), type=df.dtypes)

,unique,min,mean,50%,max,type
date,366,NaN,NaN,NaN,NaN,str
hour,24,NaN,NaN,NaN,NaN,str
passenger_count,6,NaN,NaN,NaN,NaN,str
PU_Borough,7,NaN,NaN,NaN,NaN,str
DO_Borough,7,NaN,NaN,NaN,NaN,str
payment_type,5,0.0,1.566788,1.0,4.0,int64
trip_count,4175,1.0,42.740245,3.0,7242.0,int64
trip_distance_sum,91407,0.01,143.666461,20.29,11781.52,float64
duration_sum,143340,0.016667,718.476107,68.233333,112972.7,float64
fare_amount_sum,112094,3.0,834.065682,100.0,105949.34,float64


In [9]:
# categorical variables Inspection
for col in ["date", "hour", "passenger_count", "PU_Borough", "DO_Borough", "payment_type"]:
    print(f"\n{'=' * 40}")
    print(f"{col}, unique = {df[col].nunique()}, type = {df[col].dtype}")
    print(f"{'=' * 40}")
    print(pd.Series(df[col].unique()))


date, unique = 366, type = str
0      2024-01-01
1      2024-01-02
2      2024-01-03
3      2024-01-04
4      2024-01-05
          ...    
361    2024-12-27
362    2024-12-28
363    2024-12-29
364    2024-12-30
365    2024-12-31
Length: 366, dtype: str

hour, unique = 24, type = str
0     00:00
1     01:00
2     02:00
3     03:00
4     04:00
5     05:00
6     06:00
7     07:00
8     08:00
9     09:00
10    10:00
11    11:00
12    12:00
13    13:00
14    14:00
15    15:00
16    16:00
17    17:00
18    18:00
19    19:00
20    20:00
21    21:00
22    22:00
23    23:00
dtype: str

passenger_count, unique = 6, type = str
0        1.0
1        2.0
2        3.0
3        4.0
4        5.0
5    Unknown
dtype: str

PU_Borough, unique = 7, type = str
0            Bronx
1         Brooklyn
2        Manhattan
3           Queens
4          Unknown
5    Staten Island
6              EWR
dtype: str

DO_Borough, unique = 7, type = str
0            Bronx
1        Manhattan
2         Brooklyn
3           Q

In [10]:
# correlation Inspection
variables = [
        'trip_count',
       'trip_distance_sum', 'duration_sum',
       'fare_amount_sum', 'total_amount_sum'
    ]
# Pearson: strength of the linear relationship
pearson = df[variables].corr()["trip_count"].sort_values(ascending=False)
pearson

trip_count           1.000000
total_amount_sum     0.978323
duration_sum         0.975965
fare_amount_sum      0.972087
trip_distance_sum    0.906121
Name: trip_count, dtype: float64

## **Data definition**

The DataFrame contains **921,373 observations (rows)** and **18 variables (columns)**.

Based on the number of unique values and the meaning of each variable:

| Variable                    | Type                     | Description                                                          |
| --------------------------- | ------------------------ | -------------------------------------------------------------------- |
| `date`                      | **Temporal**             | Date associated with the observation.                                |
| `hour`                      | **Categorical**          | Hour of the day associated with the observation.                     |
| `passenger_count`           | **Discrete numerical**   | Number of passengers associated with the observation.                |
| `PU_Borough`                | **Categorical**          | Pickup borough.                                                      |
| `DO_Borough`                | **Categorical**          | Drop-off borough.                                                    |
| `payment_type`              | **Categorical**          | Payment method category associated with the observation.             |
| `trip_count`                | **Discrete numerical**   | Number of trips represented by the observation.                      |
| `trip_distance_sum`         | **Continuous numerical** | Sum of trip distances represented by the observation.                |
| `duration_sum`              | **Continuous numerical** | Sum of trip durations represented by the observation.                |
| `fare_amount_sum`           | **Continuous numerical** | Sum of fare amounts represented by the observation.                  |
| `extra_sum`                 | **Continuous numerical** | Sum of extra charges represented by the observation.                 |
| `mta_tax_sum`               | **Continuous numerical** | Sum of MTA tax amounts represented by the observation.               |
| `tip_amount_sum`            | **Continuous numerical** | Sum of tip amounts represented by the observation.                   |
| `tolls_amount_sum`          | **Continuous numerical** | Sum of toll charges represented by the observation.                  |
| `improvement_surcharge_sum` | **Continuous numerical** | Sum of improvement surcharge amounts represented by the observation. |
| `congestion_surcharge_sum`  | **Continuous numerical** | Sum of congestion surcharge amounts represented by the observation.  |
| `airport_fee_sum`           | **Continuous numerical** | Sum of airport fee amounts represented by the observation.           |
| `total_amount_sum`          | **Continuous numerical** | Sum of total amounts represented by the observation.                 |

The dataset contains **one temporal variable**, **four categorical variables**, **two discrete numerical variables**, and **eleven continuous numerical variables**.

### **Categorical variables**

`date` contains **366 unique values**, ranging from **01-01-2024** to **31-12-2024**, covering the entire year 2024, including the leap day.

`hour` contains **24 unique values**, corresponding to the hours of the day.

`PU_Borough` and `DO_Borough` contain the same **7 geographical categories**: **Bronx, Brooklyn, Manhattan, Queens, Staten Island, EWR, and Unknown**.

`payment_type` contains payment categories defined as:

* 0 = Flex Fare trip
* 1 = Credit card
* 2 = Cash
* 3 = No charge
* 4 = Dispute
* 5 = Unknown
* 6 = Voided trip

However, the descriptive statistics show that the current dataset contains values only from **0 to 4**.

### **Discrete numerical variables**

`passenger_count` represents the number of passengers associated with each observation. The known values range from **1 to 5**, with an additional **Unknown** category.

`trip_count` represents the number of trips included in each observation, ranging from **1 to 7,242**, with a **median of 3 trips per observation**.

**The DataFrame does not present any missing or duplicated values, and no negative values were identified in the numerical variables.**

### **Aggregated-data consideration and grain definition**

The dataset documentation and the structure of the variables provide strong evidence that the data are **aggregated rather than individual trip-level records**. The numerical variables have different ranges, and most contain very large values compared with their mean and median. For example, the maximum `duration_sum` of 112,972.70 minutes would be unrealistic for a single trip, but it can be plausible as an aggregate value representing multiple trips. The large differences between the mean, median, and maximum also indicate that several numerical variables are **right-skewed**.

The dataset can be divided into two main parts. The first consists of **low-cardinality categorical, temporal, and discrete variables**, such as `date`, `hour`, `passenger_count`, the pickup and drop-off boroughs (`PU_Borough` and `DO_Borough`), and `payment_type`. These variables define the dimensions used to distinguish observations, and their combination uniquely identifies each row.

The second part consists of **numerical aggregate measures**, including `trip_count`, `trip_distance_sum`, `duration_sum`, `fare_amount_sum`, and `total_amount_sum`. These variables show strong positive linear correlations with `trip_count`, indicating that their values generally increase as the number of trips represented by an observation increases.

The grain is therefore **one row per unique combination of** `date`, `hour`, `passenger_count`, `PU_Borough`, `DO_Borough`, and `payment_type`, with the remaining numerical variables representing aggregated measures for that group.

In [11]:
# temporal variables Conversion
df["date"] = pd.to_datetime(df["date"])
df["hour"] = pd.to_datetime(df["hour"], format="%H:%M").dt.time

In [12]:
# categorical variables Conversion
categorical_cols = [
    "PU_Borough",
    "DO_Borough",
    "passenger_count"
]
df[categorical_cols] = df[categorical_cols].astype("category")

In [13]:
# discrete numerical variable Conversion
df["trip_count"] = df["trip_count"].astype(int)

In [14]:
# Data info Inspection
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 921373 entries, 0 to 921372
Data columns (total 18 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   date                       921373 non-null  datetime64[us]
 1   hour                       921373 non-null  object        
 2   passenger_count            921373 non-null  category      
 3   PU_Borough                 921373 non-null  category      
 4   DO_Borough                 921373 non-null  category      
 5   payment_type               921373 non-null  int64         
 6   trip_count                 921373 non-null  int64         
 7   trip_distance_sum          921373 non-null  float64       
 8   duration_sum               921373 non-null  float64       
 9   fare_amount_sum            921373 non-null  float64       
 10  extra_sum                  921373 non-null  float64       
 11  mta_tax_sum                921373 non-null  float64       
 12 

In [15]:
# Data grain Definition
grain_cols = [
    "date",
    "hour",
    "passenger_count",
    "PU_Borough",
    "DO_Borough",
    "payment_type"
]

In [16]:
# grain Validation
df.duplicated(subset=grain_cols).sum()

np.int64(0)

In [17]:
# grain Validation
df[grain_cols].nunique()

date               366
hour                24
passenger_count      6
PU_Borough           7
DO_Borough           7
payment_type         5
dtype: int64

In [18]:
# chronological sorting Inspection
df.groupby("date")["hour"].apply(lambda x: x.is_monotonic_increasing).value_counts()

hour
True    366
Name: count, dtype: int64

In [19]:
# aggregated Data Logical Consistency Inspection
calculated_total = (
    df["fare_amount_sum"]
    + df["extra_sum"]
    + df["mta_tax_sum"]
    + df["tip_amount_sum"]
    + df["tolls_amount_sum"]
    + df["improvement_surcharge_sum"]
    + df["congestion_surcharge_sum"]
    + df["airport_fee_sum"]
)

difference = calculated_total - df["total_amount_sum"]

print("Maximum absolute difference:", difference.abs().max())
print("Rows with difference > 0.01:", (difference.abs() > 0.01).sum())

Maximum absolute difference: 2.9103830456733704e-11
Rows with difference > 0.01: 0


## **Final Data Preparation Conclusion**

After converting the variables to their appropriate data types, the dataset was rechecked for **grain consistency, temporal consistency, and numerical consistency**.

The **six variables defining the observation grain** (`date`, `hour`, `passenger_count`, `PU_Borough`, `DO_Borough`, and `payment_type`) uniquely identify each row.

The **temporal checks** confirmed that the dates and hours are valid and that the hours are chronologically ordered within each day.

The **numerical consistency checks** also confirmed that the aggregate monetary components are consistent with `total_amount_sum`, with no meaningful discrepancies identified.

Together with the previous checks for **missing values, duplicates, invalid categories, and negative values**, these results indicate that the dataset is **internally consistent and ready for exploratory analysis**.

The cleaned dataset will now be exported and used as the input for the subsequent analysis notebooks.


In [20]:
# Data Export in Parquet format to preserve data types and improve storage efficiency
df.to_parquet("../data/processed/nyc_taxi_2024_clean.parquet",index=False)

In [ ]:
# checking the exported Data parquet
df_check = pd.read_parquet("../data/processed/nyc_taxi_2024_clean.parquet")
df_check.info()

<class 'pandas.DataFrame'>
RangeIndex: 921373 entries, 0 to 921372
Data columns (total 18 columns):
 #   Column                     Non-Null Count   Dtype         
---  ------                     --------------   -----         
 0   date                       921373 non-null  datetime64[us]
 1   hour                       921373 non-null  object        
 2   passenger_count            921373 non-null  category      
 3   PU_Borough                 921373 non-null  category      
 4   DO_Borough                 921373 non-null  category      
 5   payment_type               921373 non-null  int64         
 6   trip_count                 921373 non-null  int64         
 7   trip_distance_sum          921373 non-null  float64       
 8   duration_sum               921373 non-null  float64       
 9   fare_amount_sum            921373 non-null  float64       
 10  extra_sum                  921373 non-null  float64       
 11  mta_tax_sum                921373 non-null  float64       
 12 